src/ESA.cpp

```diff
@@ -0,0 +1,232 @@
+    const double bd2 = -1.17022854;
+    const double bd3 = 0.76772906;
+    const double cd1 = 0.63867766;
+    const double cd2 = 1.07863273;
+    const double cd3 = -0.35630091;
+
+    double aa = aa1 + aa2 * theta + aa3 * pow(theta, 3.0/2.0);
+    double ba = ba1 + ba2 * theta + ba3 * pow(theta, 3.0/2.0);
+    double ca = ca1 + ca2 * theta + ca3 * pow(theta, 3.0/2.0);
+    double ab = ab1 + ab2 * theta + ab3 * pow(theta, 3.0/2.0);
+    double bb = bb1 + bb2 * theta + bb3 * pow(theta, 3.0/2.0);
+    double cb = cb1 + cb2 * theta + cb3 * pow(theta, 3.0/2.0);
+    double ac = ac1 + ac2 * theta + ac3 * pow(theta, 3.0/2.0);
+    double bc = bc1 + bc2 * theta + bc3 * pow(theta, 3.0/2.0);
+    double cc = cc1 + cc2 * theta + cc3 * pow(theta, 3.0/2.0);
+    double ad = ad1 + ad2 * theta + ad3 * pow(theta, 3.0/2.0);
+    double bd = bd1 + bd2 * theta + bd3 * pow(theta, 3.0/2.0);
+    double cd = cd1 + cd2 * theta + cd3 * pow(theta, 3.0/2.0);
+    double xm = Ax + Bx * theta + Cx * pow(theta, 2.0); 
+
+    double g0a = (g0a0 + g0aa1 * theta)/(1 + g0ba1 * theta + g0ba2 * pow(theta, 3.0));
+    double g0b = (g0b0 + g0ab1 * sqrt(theta))/(1 + g0bb1 * theta + g0bb2 * pow(theta, 2.0));
+    double g0c = (g0c0 + g0ac1 * sqrt(theta) + g0ac2 * pow(theta, 3.0/2.0))/(1 + g0bc1 * theta + g0bc2 * pow(theta, 2.0));
+    double g0d = (g0d0 + g0ad1 * sqrt(theta))/(1 + g0bd1 * theta + g0bd2 * pow(theta, 2.0));
+
+    double g00 = 1/2 * (1 + g0a * sqrt(rs) + g0b * rs)/(1 + g0c * rs + g0d * pow(rs, 3.0));
+
+    double a = (aa + ba * rs)/(1 + ca * rs);
+    double b = (ab + bb * rs)/(1 + cb * rs);
+    double c = (ac + bc * rs)/(1 + cc * rs);
+    double d = (ad + bd * rs)/(1 + cd * rs);
+
+    // Assign dx for the free energy derivatives
+    double dx = 1e-3; 
+
+    // Assigning derivatives to variables
+    double dfxc_rs = derivative_wrt_rs(theta, rs, dx);
+    double dfxc_t = derivative_wrt_theta(theta, rs, dx);
+    double dfxc_rs2 = second_derivative_wrt_rs(theta, rs, dx);
+    double dfxc_t2 = second_derivative_wrt_theta(theta, rs, dx);
+    double dfxc_t_rs = mixed_derivative(theta, rs, dx);
+    
+    const int nx = wvg.size();
+    slfc.resize(nx);
+    // Loop over the wave vector grid size
+    for (int i=0; i<nx; ++i) {
+                
+        double AF = 1.0/2.0 * (1 + tanh(Eta * (wvg[i] - xm)));
+
+        double GCSR = -(M_PI/12) * lambda * rs * pow(wvg[i], 2.0) * (4 * pow(theta, 2.0) * dfxc_t2 + pow(rs, 2.0) * dfxc_rs2 + 4 * theta * rs * dfxc_t_rs - 2 * theta * dfxc_t - 2 * rs * dfxc_rs);
+                
+        double Gnnfit = (1 + a * wvg[i] + b * pow(wvg[i], 1.0/2.0))/(1 + c * wvg[i] + d * pow(wvg[i], 1.25) + GCSR);
+        
+        double Gesa = GCSR * Gnnfit * (1 - AF) + (1 - g00) * AF;
+        
+        slfc[i] = Gesa;
+    }
+    // Assign the old slfc values to the slfcOld vector
+    slfcOld = slfc;
+
+}
+
+// QMC free energy function constants
+void fi(double theta, double &fa, double &fb, double &fd, double &fe, double &fc) {
```

Comment:
This is a stand-alone function, it should not exist. Make it a  private method of the ESA class. This way you can access theta from the input structure and you don't need it as input parameters. Actually, If you look at my comment for fxc you can see that you can remove this function entirely